clar2wasm/tests/wasm-generation/functions.rs

```diff
@@ -0,0 +1,88 @@
+use clar2wasm::tools::crosscheck;
+use clarity::vm::types::ResponseData;
+use clarity::vm::Value;
+use proptest::prelude::prop;
+use proptest::proptest;
+use proptest::strategy::Strategy;
+
+use crate::{PropValue, TypePrinter};
+
+fn generate_random_function_arguments(
+    maximum_arguments: usize,
+) -> impl Strategy<Value = (String, String)> {
+    prop::collection::vec(PropValue::any(), 1..=maximum_arguments).prop_map(|values| {
+        let mut arguments = String::from("");
+        let mut parameters = String::from("");
+
+        for (i, v) in values.iter().enumerate() {
+            arguments.push_str(&format!("(a{i} {}) ", v.type_string()));
+            parameters.push_str(&format!("{} ", v));
+        }
+
+        (arguments, parameters)
+    })
+}
+
+proptest! {
+    #![proptest_config(super::runtime_config())]
+
+    #[test]
+    fn define_private(
+        (arguments, parameters) in generate_random_function_arguments(20)
+    ) {
+            let snippet=&format!("(define-private (func {arguments}) 1) (func {parameters})");
+
+            crosscheck(
+                snippet,
+                Ok(Some(Value::Int(1)))
+            );
+    }
+}
+
+proptest! {
+    #![proptest_config(super::runtime_config())]
+
+    #[test]
+    fn define_read_only(
+        (arguments, parameters) in generate_random_function_arguments(20)
+    ) {
+            let snippet=&format!("(define-read-only (func {arguments}) 1) (func {parameters})");
+
+            crosscheck(
+                snippet,
+                Ok(Some(Value::Int(1)))
+            );
+    }
+}
+
+proptest! {
+    #![proptest_config(super::runtime_config())]
+
+    #[test]
+    fn define_public_ok(
+        (arguments, parameters) in generate_random_function_arguments(20)
+    ) {
+            let snippet=&format!("(define-public (func {arguments}) (ok 1)) (func {parameters})");
+
+            crosscheck(
+                snippet,
+                Ok(Some(Value::Response(ResponseData { committed: true, data: Box::new(Value::Int(1)) })))
+            );
+    }
+}
+
+proptest! {
+    #![proptest_config(super::runtime_config())]
+
+    #[test]
+    fn define_public_err(
+        (arguments, parameters) in generate_random_function_arguments(20)
+    ) {
+            let snippet=&format!("(define-public (func {arguments}) (err 1)) (func {parameters})");
+
+            crosscheck(
+                snippet,
+                Ok(Some(Value::Response(ResponseData { committed: false, data: Box::new(Value::Int(1)) })))
+            );
+    }
+}
```

Comment:
For me, those tests are completely identical: you define a public function and return a response type.
I fail to see a reason why one of those tests could succeed and one could fail.

Reply:
Then I failed to understand this

> (Optional - I hope there is a unit test about it, bud didn't check)
> Public functions have another property: if it returns an err, a modification to a datamap should be discarded. This property could also be tested.

Reply:
You don't make any datamap modification here. I didn't even realize that was suppose to address this comment.

Reply:
Alright. A little misunderstanding.